### PR TITLE
mpsl: cx: support for nRF5340

### PR DESCRIPTION
**Pull request by jciupis**
This commit adds a MPSL_CX_PIN_FORWARDER Kconfig option that makes it
possible to forward control over Coexistence interface to the nRF5340
network core. It enables use of mpsl_cx for nRF5340 SoCs.

Signed-off-by: Jedrzej Ciupis <jedrzej.ciupis@nordicsemi.no>

### CODE REVIEW

File: subsys/mpsl/cx/Kconfig [hunk 1]
```diff
@@ -61,7 +67,7 @@ config MPSL_CX_GENERIC_3PIN
 
 endchoice # MPSL_CX_CHOICE
 
-endif	# MPSL_CX
+endif	# MPSL_CX || MPSL_CX_PIN_FORWARDER
```
Comment (ankuns) [inline subsys/mpsl/cx/Kconfig:70]: For line 61. we have select NRFX_GPIOTE, should we have this selet also for MPSL_CX_PIN_FORWARDER

Comment (jciupis) [reply subsys/mpsl/cx/Kconfig:70]: That's a good catch, thank you. I think the `select` should only work when `MPSL_CX` is enabled:
```
select NRFX_GPIOTE if MPSL_CX
```
What do you think?

**Review by ankuns — APPROVED** [review]

File: subsys/mpsl/cx/thread/mpsl_cx_thread.c [hunk 2]
```diff
@@ -11,7 +11,10 @@
  *
  */
 
+
+#if !defined(CONFIG_MPSL_CX_PIN_FORWARDER)
```
Comment (hubertmis) [inline subsys/mpsl/cx/thread/mpsl_cx_thread.c:15]: I find confusing reusing the same module as "mpsl_cx_pin_forwarder" and "mpsl_cx_thread" implementation. The responsibility of the modules is different, only small part of the source file is shared, most is excluded by preprocessor macros. I think it would be cleaner to introduce a new module like `subsys/mpsl/cx/pin_forwarder.c` for this purpose.

File: subsys/mpsl/cx/thread/mpsl_cx_thread.c [hunk 3]
```diff
@@ -22,8 +25,8 @@
 
 #include "hal/nrf_gpio.h"
 
-#if DT_NODE_HAS_STATUS(DT_PHANDLE(DT_NODELABEL(radio), coex), okay)
-#define CX_NODE DT_PHANDLE(DT_NODELABEL(radio), coex)
+#if DT_NODE_EXISTS(DT_NODELABEL(nrf_radio_coex))
```
Comment (hubertmis) [inline subsys/mpsl/cx/thread/mpsl_cx_thread.c:28]: Also, I find it confusing that "pin forwarder" and actual coex implementation share the same DTS description. Pin forwarder should simply use description in DTS which pins are to be connected to Net Core, while coex needs to know details that given pins are connected to external PTA, which pin is in/out, active high/low, pin role, etc.

**Review by hubertmis — COMMENTED** [review]
I'm not a fan of mixing "pin forwarding" functionality with features actually using the forwarding pins. For both device tree and implementation. Am I the only one who finds it hard to follow?

File: subsys/mpsl/cx/thread/mpsl_cx_thread.c (hunk 2 above)
Comment (ankuns) [reply subsys/mpsl/cx/thread/mpsl_cx_thread.c:15]: I don't agree. Consider you are customer, not developer of NCS. You want to add your custom implementation, howewer you pull NCS with west into your project. You don't want to have to tweak anything inside the NCS itself. You want to add another cx implementation. If pin_forwarder functionality is crammed into single "pin_forwarder.c" you will need to tweak this file to add your new cx plug-in. You cannot add new cx implementation seamlessly.
Let me remind you this PR https://github.com/nrfconnect/sdk-nrf/pull/6502

I think that this "pin-forwarding" feature is small enough that it can be kept within implementation.

File: subsys/mpsl/cx/thread/mpsl_cx_thread.c (hunk 3 above)
Comment (ankuns) [reply subsys/mpsl/cx/thread/mpsl_cx_thread.c:28]: Generally I agree with you. This should be designed better, app should have list of pins in dts to forward, net core should simply use them.  However we are not sure how to solve pin forwarding in both elegant and easy for end user way. Consider `-DSHIELD` , consider overlays (for end user convenience possibly having just one overlay file per feature like cx should be considered), consider adding next pins to forward as a result of using features like "cx", fem" and whatever we need and what does not know about each other.
The case is also that we need this just now, we can open later other PR when we find way for better pin forwarding.

Comment (jciupis) [reply subsys/mpsl/cx/thread/mpsl_cx_thread.c:28]: I agree that the information both cores need from the devicetree description of Coex node differs and that the application core could do with pin numbers only. However, the approach where only a single .dts file is needed felt simpler to me. For example, a single .dts file describing a Coex shield could be applied to both cores. IMO it makes sense for a single shield to have a single devicetree description.

I'm aware that the entire concept of pin forwarder handling the pin assignments is somewhat crooked and should probably be handled in an entirely different manner. I don't think that having separate DTS descriptions changes the situation much - the solution is still not very elegant, but requires more effort to be used with shields conveniently.

Comment (hubertmis) [reply subsys/mpsl/cx/thread/mpsl_cx_thread.c:28]: Using separate DTS node for describing pin forwarding does not prevent us from using single DTS overlay for a shield. I was thinking about something like:

```
// DTS shield overlay describing a part of external radio chip

&pin_forwarder {
  coex_pins = <&gpio0 7 &gpio1 8 &gpio1 12>;
  irq_pin = <&gpio1 11>
};

/ {
  radio_coex: radio_coex_three_wire {
    status = "okay";
    compatible = "generic-radio-coex-three-wire";
    req-gpios =     <&gpio0 7 (GPIO_PULL_DOWN | GPIO_ACTIVE_HIGH)>;
    pri-dir-gpios = <&gpio1 8 (GPIO_PULL_UP   | GPIO_ACTIVE_HIGH)>;
    grant-gpios =   <&gpio1 12 (GPIO_PULL_DOWN | GPIO_ACTIVE_HIGH)>;
  };
};

&radio {
  coex = <&radio_coex>;
};
```

This way we explicitly state which pins are going to be forwarded by App Core, instead of implicitly passing this through `generic-radio-coex-three-wire` node which should not be parsed by App Core at all.

Comment (jciupis) [reply subsys/mpsl/cx/thread/mpsl_cx_thread.c:28]: > Net Core would ignore pin_forwarder node, while App Core ignores radio_coex and radio

I'm not aware of a possibility to ignore parts of an overlay. AFAIK, an attempt to apply the above overlay to the application core will result in a build error because of a non-existing node `radio` being configured. Probably `pin_forwarder` would yield a similar error when applying that overlay to the network core.

File: subsys/mpsl/cx/thread/mpsl_cx_thread.c (hunk 2 above)
Comment (hubertmis) [reply subsys/mpsl/cx/thread/mpsl_cx_thread.c:15]: I disagree :)

While I'm creating my out-of-tree application for NCS I need an easy and clear way to define my boards and configuration for images for all cores. Also I would appreciate if creating new FEM or cx implementations was easy.

I would like pin forwarding to be handled by the system based on device tree only, without need for changes in any implementation like cx or FEM. If there was a generic pin-forwarder feature like I described in https://github.com/nrfconnect/sdk-nrf/pull/6706#discussion_r796387363, I would need to place appropriate `pin_forwarding` node in DTS file I prepare for my App Core image. And that's it. I would not need to change anything in any implementation related to coexistence or FEMs.

If I use any of the cx implementation provided in SDK, I simply add them to device tree for my Net Core. And that's it.

When I create a new cx implementation, I focus only on part related to coexistence itself, the one needed for Net Core. I do not want to implement pin forwarding in this file. With the patch I'm reviewing it's not the case. I would need to mix pin forwarding with new cx implementation I'm creating. I need to implement both pin forwarding feature and cx functionality in single source file, what is pretty hard to follow if you don't know SDK internals.

That's why I would suggest to separate pin forwarding from FEMs or cx implementations and make it explicit and easy to use.

File: subsys/mpsl/cx/thread/mpsl_cx_thread.c (hunk 3 above)
Comment (hubertmis) [reply subsys/mpsl/cx/thread/mpsl_cx_thread.c:28]: Yeah, probably should be described a little differently, like:
```
/{
  pin_forwarder {
    ...
  };

  radio {
    ...
  };
};
```

instead of using references. But it does not change the essence of my proposal

File: subsys/mpsl/cx/thread/mpsl_cx_thread.c (hunk 2 above)
Comment (jciupis) [reply subsys/mpsl/cx/thread/mpsl_cx_thread.c:15]: @hubertmis I like your proposal a lot, I think it's easy to use and elegant. The only problem I see with that approach is that I don't know yet how to implement it :) The `pin_forwarder` node would have to allow the user to provide an unspecified number of pins to forward and I'm not sure how to represent it in devicetree. I'll see what I can do.

Comment (hubertmis) [reply subsys/mpsl/cx/thread/mpsl_cx_thread.c:15]: I'm pretty sure it's possible to iterate over child nodes in device tree. Let's dig into this to find the right solution!

**Comment by jciupis** [comment]
Summary of F2F discussions:
The plan is to implement a generic pin forwarding solution according to @hubertmis's proposal and contribute it upstream. Since that is not feasible within the time frame of NCS 1.9.0 we're going to proceed with this PR now and replace it with the generic solution once it's reviewed & merged upstream.

**Review by hubertmis — APPROVED** [review]
Approving, but I hope a generic solution will replace this soon

**Comment by martintv** [comment]
approving

**Review by martintv — COMMENTED** [review]
approving
